src/metax_api/tests/api/rest/base/views/datasets/write.py

```diff
@@ -3899,6 +3914,97 @@ def test_user_projects_are_checked_when_writing_into_protected_data_catalog(self
+        response = self.client.post('/rest/datasets', self.cr_test_data, format="json")
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED, response.data)
+        self.assertEqual(response.data['research_dataset']['preferred_identifier'], '123456')
+
+    def test_external_service_can_update_cr_in_own_catalog(self):
+        self.cr_test_data['data_catalog'] = EXT_CATALOG
+        self.cr_test_data['research_dataset']['preferred_identifier'] = '123456'
+        response = self.client.post('/rest/datasets', self.cr_test_data, format="json")
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED, response.data)
+        self.assertEqual(response.data['research_dataset']['preferred_identifier'], '123456')
+
+        cr_id = response.data['id']
+        self.cr_test_data['research_dataset']['preferred_identifier'] = '654321'
+        response = self.client.put('/rest/datasets/{}'.format(cr_id), self.cr_test_data, format="json")
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK, response.data)
+        self.assertEqual(response.data['research_dataset']['preferred_identifier'], '654321')
+
+    def test_external_service_can_not_add_cr_to_other_catalog(self):
+        dc = self._get_object_from_test_data('datacatalog', requested_index=1)
+        self.cr_test_data['data_catalog'] = dc['catalog_json']['identifier']
+        response = self.client.post('/rest/datasets', self.cr_test_data, format="json")
+
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN, response.data)
+
+    def test_external_service_can_not_update_cr_in_other_catalog(self):
+        self._use_http_authorization()
+        dc = self._get_object_from_test_data('datacatalog', requested_index=1)
+        self.cr_test_data['data_catalog'] = dc['catalog_json']['identifier']
+        response = self.client.post('/rest/datasets', self.cr_test_data, format="json")
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED, response.data)
+
+        cr_id = response.data['id']
+        self._use_http_authorization(username=django_settings.API_EXT_USER['username'],
+            password=django_settings.API_EXT_USER['password'])
+        self.cr_test_data['research_dataset']['preferred_identifier'] = '654321'
+        response = self.client.put('/rest/datasets/{}'.format(cr_id), self.cr_test_data, format="json")
+
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN, response.data)
+
+    def test_external_service_not_creating_cr_identifier(self):
+        self.cr_test_data['data_catalog'] = EXT_CATALOG
+
+        response = self.client.post('/rest/datasets', self.cr_test_data, format="json")
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED, response.data)
+        self.assertIsNone(response.data['research_dataset'].get('preferred_identifier'))
+        self.assertIsNotNone(response.data['research_dataset']['access_rights'])
```

Comment:
I actually think that empty ``preferred_identifier`` should never be allowed. If current implementation allows it, then it is probably an existing bug and should be prevented. Checking what should be done in that situation...

Reply:
do you mean empty `preferred_identifier` is 
`{'preferred_identifier':''}` 
or completely missing?

Reply:
Both, ``cr.research_dataset.preferred_identifier`` should never be missing or be null or blank string once a dataset has been created into Metax. This is a bug in Metax that seems to have existed for quite a while now already. Luckily there are zero datasets that have suffered from this.